pytential/source.py

```diff
@@ -128,10 +128,16 @@ def ambient_dim(self):
         return self._nodes.shape[0]
 
     def op_group_features(self, expr):
-        from sumpy.kernel import AxisTargetDerivativeRemover
+        from pytential.utils import sort_arrays_together
+        # since IntGs with the same source kernels and densities calculations
+        # for P2E and E2E are the same and only differs in E2P depending on the
+        # target kernel, we group all IntGs with same source kernels and densities.
+        # sorting is done to avoid duplicates as the order of the sum of source
+        # kernels does not matter.
         result = (
-                expr.source, expr.densities, expr.source_kernels,
-                AxisTargetDerivativeRemover()(expr.target_kernel),
+                expr.source,
+                *sort_arrays_together(expr.densities, expr.source_kernels),
```

Comment:
I'm honestly surprised this succeeds---pymbolic expressions are [unordered](https://github.com/inducer/pymbolic/blob/ca9dcc5befe13b7a29fdfbc21ec687a23398b2ec/pymbolic/primitives.py#L633-L643)? What is it sorting by? [For the target derivatives](https://github.com/inducer/pytential/blob/b6e627ef815c740324e553615b2f4e21292b39ae/pytential/symbolic/compiler.py#L577), I ended up using `repr` as the sort key.